basic_pipeline/8_Mixer.md

```diff
@@ -0,0 +1,232 @@
+  }}
+ end
+ ...
+end
+```
+We have provided an `handle_init/1` callback, which does not expect any options to be passed. We are simply setting up the structure of the element state.
+As mentioned previously, we will have a `Track` structure for each of the input pads.
+Following on the callbacks implementation, let's continue with `handle_process/4` implementation:
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ @impl true
+ def handle_process(pad, buffer, _context, state) do
+  tracks =
+    Map.update!(state.tracks, pad, fn track ->
+      %Track{track | buffer: buffer}
+    end)
+
+  state = %{state | tracks: tracks}
+  {{:ok, [{:redemand, :output}]}, state}
+ end
+ ...
+end
+```
+What we do is that we are simply putting the incoming `buffer` into the `Track` structure for the given pad - note, that we are sure that we are not losing any information which is in the `Track`'s buffer before the update. Why can we be sure of that? That's because we will be returning the `:redemand` action. As you might remember from the chapter about the redemands, all the actions returned from the element's callback are taken immediately after the callback returns - that means, they are taken before any other incoming event is served. If we would have put the whole logic of processing the buffer into the `handle_demand/5` callback, we could simply call that callback with the `:redemand` action and be sure that the newly added buffer will be immediately processed before any other buffer comes (that means - before the other `handle_process/4` callback gets called).
+Such an approach is powerful since we can also use it in other callback - `handle_end_of_stream/3`, which get's called once the `:end_of_stream` event comes on the given pad.
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ @impl true
+ def handle_end_of_stream(pad, _context, state) do
+  tracks =
+    Map.update!(state.tracks, pad, fn track ->
+      %Track{track | status: :finished}
+    end)
+
+  state = %{state | tracks: tracks}
+  {{:ok, [{:redemand, :output}]}, state}
+ end
+ ...
+end
+```
+
+What we did here was similar to the logic defined in the `handle_process/4` - we have just updated the state of the track (in that case - by setting its status as `:finished`) and then we called the `handle_demand/5` callback with the `:redemand` actions. The `handle_demand/5` will take care of the fact that the track state has changed.
+There is nothing left to do apart from defining the `handle_demand/5` itself!
+```Elixir
+# FILE: lib/elements/Mixer.ex
+
+defmodule Basic.Elements.Mixer do
+ ...
+ @impl true
+ def handle_demand(:output, size, _unit, ctx, state) when size > 0 do
+  {state, actions} = update_state_and_prepare_actions(state, ctx.pads)
+  {{:ok, actions}, state}
+ end
+
+ @impl true
+ def handle_demand(_ref, _size, _unit, _ctx, state), do: {state, []}
+ ...
```

Comment:
Should it be `...` here?

Reply:
Probably it should be - I was trying to point out that there will be something else in the Basic.Elements.Mixer implementation